// Copyright 2023 Intrinsic Innovation LLC

syntax = "proto3";

package intrinsic_proto.skills;

import "google/protobuf/empty.proto";
import "intrinsic/skills/proto/skills.proto";

message GetSkillsResponse {
  // The returned skills. The contents of this field will only be valid when
  // `status` is OK.
  repeated intrinsic_proto.skills.Skill skills = 2;

  reserved 1;
  reserved "status";
}

message GetSkillRequest {
  // The identifier for the skill in the form `<package>.<name>`.
  //
  // This should be something like ai.intrinsic.hello_world
  string id = 2;

  reserved 1;
  reserved "skill_name";
}

message GetSkillResponse {
  // Information about the Skill.
  intrinsic_proto.skills.Skill skill = 1;
}

message ListSkillsRequest {
  // Upper limit for the returned page size, i.e. the service might return fewer
  // results.
  //
  // Default behavior: returns a page size of 20. A page_size of 0 will result
  // in default behavior.
  //
  // Maximum value is 200. If this is exceeded, returns 200.
  //
  // Negative values are invalid and will result in an INVALID_ARGUMENT error.
  int32 page_size = 1;

  // If the page token is empty, up to `page_size` first elements of the
  // collection will be returned. On subsequent requests, set the
  // `next_page_token` obtained from the previous call to retrieve the next
  // page.
  string page_token = 2;

  // A filter expression to filter the returned skills.
  //
  // The format of the filter expression follows the specification outlined in
  // https://google.aip.dev/160. Only the following field are supported:
  //   * `sideloaded`
  string filter = 3;
}

message ListSkillsResponse {
  // Skills matching the request.
  repeated intrinsic_proto.skills.Skill skills = 1;

  // Pass this token to the subsequent list requests in to obtain the next page.
  string next_page_token = 2;
}

service SkillRegistry {

  // Deprecated: Use `ListSkills` instead. Important: `ListSkills` is always
  //   paginated. If you require all skills in the registry (behavior of
  //   `GetSkills`) paginate through all pages.
  rpc GetSkills(google.protobuf.Empty) returns (GetSkillsResponse) {
    option deprecated = true;
  }

  // Returns information about the requested Skill.
  rpc GetSkill(GetSkillRequest) returns (GetSkillResponse) {}

  // Lists all of the skills currently installed.
  rpc ListSkills(ListSkillsRequest) returns (ListSkillsResponse) {}
}
